File: Manufacturing/src/Python/package/microsoft_azure_sphere_deviceapi/sideload.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os
from typing import Literal

import microsoft_azure_sphere_deviceapi.utils as utils
from microsoft_azure_sphere_deviceapi.exceptions import ValidationError

__all__ = ['delete_image', 'install_images', 'stage_image']


def delete_image(component_id: str) -> dict:
    """Makes a "DELETE" request to remove a component from the device.
    This requires enabling development mode.

    :param component_id: The component id of the image to be deleted.
    :type component_id: string
    :return: An empty response on success. An exception will be thrown on error.
    :rtype: Tuple(int, dict[str, str])
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """

    if not component_id or not utils.is_uuid(component_id):
        raise ValidationError(
            "ERROR: Cannot delete image, invalid component ID.")

    return utils.delete_request(f"app/image/{component_id}")


def install_images(app_control_mode: Literal["Auto", "Manual"] = "Auto") -> dict:
    """Makes a "POST" request to install all staged images on a device.
    For deploying unsigned images, this requires enabling development mode.

    :param app_control_mode: Determines if application starts automatically after installation,
    defaults to 'Auto'.
    :type app_control_mode: str
    :accepted-values app_control_mode: Auto, Manual
    :return: An empty response on success. An exception will be thrown on error.
    :rtype: Tuple(int, dict[str, str])
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """

    valid_app_control_modes = ["Auto", "Manual"]

    if app_control_mode not in valid_app_control_modes:
        raise ValidationError(
            "ERROR: Cannot set the app status, invalid component ID.")

    return utils.post_request("update/install", {"appControlMode": app_control_mode})


def stage_image(image_path: str) -> dict:
    """Makes a "PUT" request to stage an image on a device.
    For deploying unsigned images, this requires enabling development mode.

    :param image_path: The absolute path to an image package.
    :return: An empty response on success. An exception will be thrown on error.
    :rtype: Tuple(int, dict[str, str])
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """

    if not image_path:
        raise ValidationError(
            "ERROR: Cannot stage image, image path is null or empty.")

    if not os.path.isfile(image_path):
        raise ValidationError(
            "ERROR: Cannot stage image, image path is null or empty.")

    with open(image_path, "rb") as opened_image:
        return utils.put_request_octet_stream(
            "update/stage", opened_image.read()
        )

File: Manufacturing/src/CSharp/Nuget/Package/Sideload.cs
/* Copyright (c) Microsoft Corporation. All rights reserved.
   Licensed under the MIT License. */
namespace Microsoft.Azure.Sphere.DeviceAPI
{
    using System;
    using System.IO;

    /// <summary>
    /// Device REST APIs to deploy and manage applications on the attached device.
    /// </summary>
    public static class Sideload
    {
        /// <summary>Makes a "DELETE" request to remove a component from an attached device. This requires enabling development mode.</summary>
        /// <param name="componentID">The component id of the image to be deleted.</param>
        /// <returns>An empty response as a string on success. An exception will be thrown on error.</returns>
        public static string DeleteImage(string componentID)
        {
            if (string.IsNullOrEmpty(componentID) || !Validation.IsUuid(componentID))
            {
                throw new ValidationError("Cannot delete image, invalid component ID.");
            }

            return RestUtils.DeleteRequest($"app/image/{componentID}");
        }

        /// <summary>Makes a "POST" request to install all staged images on an attached device. For deploying unsigned images, this requires enabling development mode.</summary>
        /// <param name="appControlMode">Determines if application starts automatically after installation, defaults to 'Auto'.</param>
        /// <remarks>Accepted trigger values: ["Auto", "Manual"]</remarks>
        /// <returns>An empty response as a string on success. An exception will be thrown on error.</returns>
        public static string InstallImages(string appControlMode = "Auto")
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("InstallImages", "3.0.0");
            string[] validAppControlModes = new string[] { "Auto", "Manual" };

            if (string.IsNullOrEmpty(appControlMode) || !Array.Exists(validAppControlModes, elem => elem.Equals(appControlMode)))
            {
                throw new ValidationError("Cannot install image, app control mode is invalid.");
            }

            return RestUtils.PostRequest("update/install", new { appControlMode });
        }

        /// <summary>Makes a "PUT" request to stage an image on an attached device. For deploying unsigned images, this requires enabling development mode.</summary>
        /// <param name="imageLocation">The file location of the image you would like to stage on an attached device.</param>
        /// <returns>An empty response as a string on success. An exception will be thrown on error.</returns>
        public static string StageImage(string imageLocation)
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("StageImage", "3.0.0");
            if (string.IsNullOrEmpty(imageLocation))
            {
                throw new ValidationError("Cannot stage image, image location is null or empty.");
            }

            if (!File.Exists(imageLocation))
            {
                throw new ValidationError("Cannot stage image, image location does not exist or is not accessible.");
            }

            byte[] imagePackageBytes = File.ReadAllBytes(imageLocation);

            return RestUtils.PutWithOctetStreamRequest("update/stage", imagePackageBytes);
        }
    }
}
